### PR TITLE
LPS-123211 Fixes failed navigation on CORS redirect issues

### PR DESCRIPTION
**Pull request by jbalsas**
As outlined in both [Error message shows when editing the checkouted document in Sharepoint REST repository](https://issues.liferay.com/browse/LPS-123214) and [An error message will be displayed when adding google doc/slide/sheet](https://issues.liferay.com/browse/LPS-123211), it's currently not possible to access those services.

**Disclaimer**
I've only been able to properly setup and test the GDrive scenario and derive from there that any impediment this was causing should also be gone for the Sharepoint scenario (which is harder to configure).

**What's the problem**
Redirects to different domains done through `XMLHttpRequest` or `fetch` failed due to CORS restrictions.

**The fix**
More than a fix, a workaround, we already had in place a condition by which a "premature termination" would force a manual retry of the navigation via `window.location = event.path`. This handles the redirect directly from the server avoiding the CORS issue.

When migrating from `metal-ajax` to `fetch`,  the condition no longer holds true since the error is now "Failed to fetch". This could probably be improved further, but to simplify the fix, we simply add that to the set of conditions that can trigger a manual retry. 

If you try this scenario (now and even before in 7.3.x), you should be able to see the failed xhr attempt in the console if you preserve the logs and then the manual retry, so this should keep the existing behaviour.

**The future**
We should investigate further if a better approach that works through `fetch` is possible.

### CODE REVIEW

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
Cool thing is we can get this fixed directly in DXP without having to cut a senna release 😂

**Comment by jbalsas** [comment]
ci:test:relevant

**Review by markocikos — APPROVED** [review]
It seems a bit weird to switch over error message and not something like code or type. But, PR looks the be solving the issue consistently with existing code, so :+1:

**Comment by jbalsas** [comment]
> It seems a bit weird to switch over error message and not something like code or type. But, PR looks the be solving the issue consistently with existing code, so 👍

You're totally right... I think this was the only way we found back in the day to deal with this:
- [add functionality to catch CORS errors and attempt to forward the user to the desired link with no use of XHR](https://github.com/liferay/senna.js/commit/e2523944a21b7dfba7286b3c2b09cb4e5f727c4a)
- [enhance detection of CORS issue and replace it with a more generic na](https://github.com/liferay/senna.js/commit/e1cd6b5d474a60873f6f882393c211b75279e6f4)

As I state, I think we might want to rethink many things in here, but the first step is to get it working.

**Comment by wincent** [comment]
My only concern here is whether the error text might vary by browser. Chrome, Firefox, even Safari...

**Comment by jbalsas** [comment]
> My only concern here is whether the error text might vary by browser. Chrome, Firefox, even Safari...

Ugh, you're right, I was in the middle of checking it and forgot because I couldn't find the definition of this in the spec 😢 

Firefox in fact appears to throw `TypeError: NetworkError when attempting to fetch resource`

**Comment by wincent** [comment]
> > My only concern here is whether the error text might vary by browser. Chrome, Firefox, even Safari...
> 
> Ugh, you're right, I was in the middle of checking it and forgot because I couldn't find the definition of this in the spec 😢
> 
> Firefox in fact appears to throw `TypeError: NetworkError when attempting to fetch resource`

Nothing that can't be solved with a regex...

regex

... or something 🤦‍♂️

**Comment by jbalsas** [comment]
So...

**Chrome:** TypeError: Failed to fetch
**Firefox:** TypeError: NetworkError when attempting to fetch resource
**Safari:** TypeError: Preflight response is not successful

The good thing is at least they don't change if you change the browser language :D 

I think we can simply get rid of the switch and always reattempt the manual navigation if something failed. Status codes are dealt in the `then` through the `assertValidStatusCode` so a 404 or a 500 won't trigger a refresh which would be reserved for only the more severe and esoteric `TypeError` issues.

How does that sound?

**Comment by wincent** [comment]
> I think we can simply get rid of the switch and always reattempt the manual navigation if something failed. Status codes are dealt in the `then` through the `assertValidStatusCode` so a 404 or a 500 won't trigger a refresh which would be reserved for only the more severe and esoteric `TypeError` issues.
> 
> How does that sound?

A bit scary, as it's a non-minimal change that _might_ have some unpleasant consequences (I don't know though).

Having said that, a middle ground would be to:

1. Add new cases that you've documented, preserving the existing special case for timeouts (to keep change minimal).
2. In the `default`, when `process.env.NODE_ENV === 'development'` log whenever we see an unexpected message, so that we at least have a chance of noticing it when testing locally or in CI.
3. Additionally, do what you said (attempt force reload) for `default` case.

But if you want to go for the more aggressive change, I'm fine with that too. Who doesn't like a bit of gambling? In the end it's hopefully never any worse than running with SPA disabled.

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
Hey @wincent, I ended up just adding both Firefox (ending in `.`) and Safari error messages to the `premature` condition and kept the rest all the same (not even logging additional issues since the default case also covers `INVALID_STATUS` codes, so that would probably get noisy.

**Comment by jbalsas** [comment]
Going to forward so we can get this out of the way since it's a possible blocker for 7.3 FP1 but happy to follow up if needed!

**Comment by jbalsas** [comment]
ci:forward

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/RequestScreen.js [hunk 1]
```diff
@@ -282,20 +286,23 @@ class RequestScreen extends Screen {
...
 			switch (reason.message) {
-				case REQUEST_TIMEOUT:
+				case REQUEST_TIMEOUT_MSG:
 					reason.timeout = true;
 					break;
-				case REQUEST_PREMATURE_TERMINATION:
+				case REQUEST_PREMATURE_TERMINATION_MSG:
+				case FAILED_TO_FETCH_MSG:
+				case NETWORK_ERROR_MSG:
+				case PREFLIGHT_ERROR_MSG:
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/RequestScreen.js:301]: I'm amazed you got away without sorting this.

Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/RequestScreen.js:301]: I swear I sorted it... in some version of the PR 😂

**Pull request closed without merging**